Jenkinsfile for Azure

colemickens opened the pull request:
This adds a basic Jenkinsfile for testing kubernetes-anywhere (+kubernetes) on Azure.

Currently it builds HEAD of master of `kubernetes` and tests that along with the "current" branch of `kubernetes-anywhere` checked out by the Jenkins job. For now this will be master on a regular interval, with PRs being built on demand and tested as well.

Goal is to provide some basic testing of k-a and kubernetes, and allow me to iterate on some other PRs faster (adding back premium storage, ssd support, isolated deploy dirs).

This actually doesn't work on `master` of Kubernetes due to three issues:

1. (general issue) Problems handling paths with spaces. [PR for this issue.](https://github.com/kubernetes/kubernetes/pull/30583)

2. (recent master regression) [kubelet fails to launch static pods: "inspected image [] does not match []" ](https://github.com/kubernetes/kubernetes/issues/30580). [PR for this issue.](https://github.com/kubernetes/kubernetes/pull/30582)

2. (recent master regression) [apiserver doesn't come up anymore: "can't setup API"]()

Any help on getting the first two merged and the third one investigated is appreciated.

In the future there will be another Jenkinsfile for building master/prs of kubernetes and then testing that against the last release and master of kubernetes-anywhere. This will also be a jenkins-dsl job used for reconstituting all of these jobs on a new Jenkins server, so that this work is easily reproducible and so that others can setup CD for Azure.

Future work:
* Another Jenkinsfile oriented around building `kubernetes` master/pull-requests using the latest stable of `kubernetes-anywhere`
* job-dsl definition for all of the Azure Jenkins job so it can easily be re-setup or so other users can run Azure Kubernetes CD.
* Run conformance tests (need help from @mikedanese on this)

Note, this also includes getting Azure over onto Ignition.

This is currently running at https://jenkins.polykube.io. The following GitHub users have been granted access: @brendandburns @rootfs @mikedanese

<img width="859" alt="screen shot 2016-08-15 at 11 25 24 am" src="https://cloud.githubusercontent.com/assets/327028/17676907/5bd4a590-62e5-11e6-9e1a-b0f69e94609d.png">

colemickens commented:
This also fixes a weird issue with addons where heapster can take `10.0.0.10` and keep kube-dns from working. Strange that I managed to hit it, but I did.

mikedanese commented:
@colemickens does that actually fix anything? Does it change the order that the addons are applied?

colemickens commented:
@mikedanese Yes, it just re-orders them so that `kube-dns` service gets deployed first and takes `10.0.0.10`, ensuring that `heapster` can't accidentally take it.

mikedanese commented:
I don't understand this causes the services to be deployed in a different order. They still get written to .tmp directory and are kubectl applied?

colemickens commented:
Sorry @mikedanese, you're right. For some reason I thought they were stuck together in a List object or something. I'll drop that commit, we'll have to fix it by renaming the files so they execute in a pre-ordained order or by specifying a fixed service ip for `heapster` and any other Services in addons.

mikedanese commented:
sure thing

mikedanese commented on phase1/azure/Kconfig:
```diff
@@ -51,7 +51,7 @@ config phase1.azure.admin_username
 config phase1.azure.admin_password
 	string "Virtual Machine Admin Password"
-	default "AzureKubernet3s!"
+	default "Ch@n3ThisPassword"
```
missing a g in change

mikedanese commented:
LGTM. would be good to not pull /srv/kubernetes/apiserver-key.pem onto nodes eventually.

mikedanese commented:
Ignition was fairly easy to switch to?

colemickens commented:
I just fixed the password (good catch!) and removed some old debug echos and rebased it.

>LGTM. would be good to not pull /srv/kubernetes/apiserver-key.pem onto nodes eventually.

Agreed. It's on the list somewhere... might even be an Issue already.

>Ignition was fairly easy to switch to?

Very. All I did was update those ~5 lines in configure-vm.sh. The changes for the ephemeral drive looked to be very painless as well.

colemickens commented:
By the way, `kubernetes` `master` was fixed enough that the job is now running successfully. I've granted everyone in the `Microsoft` and `Azure` orgs access, as well as `mikedanese`. I've opted not to open the Jenkins instance to all of `kubernetes` organization because of creds that are exposed in the job details.

https://jenkins.polykube.io/job/kubernetes-anywhere-azure_master/

<img width="1004" alt="screen shot 2016-08-17 at 2 02 12 pm" src="https://cloud.githubusercontent.com/assets/327028/17753037/3d7d6f9c-6483-11e6-98ec-60fce69d5227.png">
